### PR TITLE
storage: write source data to persist in batches

### PR DESCRIPTION
**Pull request by guswynn**
Closes https://github.com/MaterializeInc/materialize/issues/13116

As @benesch said: "We don't need to get too fancy with the batch sizing policy to start; anything rudimentary that bounds memory usage would be acceptable.". This pr implements the most basic batching policy: we make a new batch _per-timestamp_, and use `100` as the `size_hint`. Its possible that if we KNOW a batch is empty, we can skip that batch. Its unclear to me if that still incurs an additional s3 write (cc @danhhz )

### Motivation
  * This PR adds a known-desirable feature.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
@philip-stoev may want to run zippy against this

### Release notes

None

### CODE REVIEW

File: src/storage/src/render/persist_sink.rs [hunk 1]
```diff
@@ -89,45 +90,81 @@ pub fn render<G>(
...
+                        // TODO(aljoscha): Figure out how errors from this should be reported.
+
+                        let expected_upper = shared_frontier.borrow().clone();
+
+                        let mut batch = stashed_batches
+                            .remove(&ts)
+                            .expect("batch for timestamp to still be there")
+                            .finish(input_frontier.clone())
+                            .await
+                            .expect("invalid usage");
```
Comment (guswynn) [inline src/storage/src/render/persist_sink.rs:148]: @danhhz  if the write client provides an easy way to make an "empty" `Batch`, then these 2 calls can be consolidated and this code will be easier to read!

File: src/storage/src/render/persist_sink.rs [hunk 2]
```diff
@@ -89,45 +90,81 @@ pub fn render<G>(
                 while let Some((_cap, data)) = input.next() {
                     data.swap(&mut buffer);
 
+                    let mut last_frontier = shared_frontier.borrow().clone();
+                    // TODO: come up with a better default batch size here
```
Comment (danhhz) [inline src/storage/src/render/persist_sink.rs:94]: the hint just pre-sizes some internal buffers (think Vec::with_capacity), the size of the chunks is controlled by https://github.com/MaterializeInc/materialize/blob/173a1316a67c73093debcc8892a0dfde7ace7415/src/persist-client/src/lib.rs#L183. if you don't have a good estimate, it's perfectly fine to pass 0. 100 also works

File: src/storage/src/render/persist_sink.rs (hunk 1 above)
Comment (danhhz) [reply src/storage/src/render/persist_sink.rs:148]: oh it's safe to call compare_and_append_batch after not adding any updates to it. it already is smart about not writing anything to s3 if you haven't added data to the builder

File: src/storage/src/render/persist_sink.rs [hunk 3]
```diff
@@ -89,45 +90,81 @@ pub fn render<G>(
...
+                            .expect("batch for timestamp to still be there")
+                            .finish(input_frontier.clone())
+                            .await
+                            .expect("invalid usage");
+
+                        write
+                            .compare_and_append_batch(
+                                &mut batch,
+                                expected_upper,
+                                input_frontier.clone(),
```
Comment (guswynn) [inline src/storage/src/render/persist_sink.rs:154]: oh i think these are wrong, we need to be able to compare and append multiple things

File: src/storage/src/render/persist_sink.rs [hunk 4]
```diff
@@ -89,45 +89,94 @@ pub fn render<G>(
...
-
+                // See if any timestamps are done!
+                // TODO(guswynn/petrosagg): remove this additional allocation
+                let mut finalized_timestamps: Vec<_> = stashed_batches
+                    .keys()
+                    .filter(|ts| !input_frontier.less_equal(ts))
+                    .copied()
+                    .collect();
+                finalized_timestamps.sort_unstable();
+                let finalized_batch_count = finalized_timestamps.len();
```
Comment (aljoscha) [inline src/storage/src/render/persist_sink.rs:118]: nit: could move this one right in front of the loop where you use it.

File: src/storage/src/render/persist_sink.rs (hunk 2 above)
Comment (aljoscha) [reply src/storage/src/render/persist_sink.rs:92]: We could make the size hint adaptive, by observing what batch sizes we end up writing, dynamically. That's an over-optimization for now, though.

**Review by aljoscha — APPROVED** [review]

File: src/storage/src/render/persist_sink.rs (hunk 2 above)
Comment (guswynn) [reply src/storage/src/render/persist_sink.rs:92]: @petrosagg and I have some plans on how to improve this impl quite a bit, so I will leave this for now

**Comment by philip-stoev** [comment]
Full Nightly, including Zippy, is scheduled.

**Pull request merged**